new python backend

gw000 opened the pull request.

dhatch commented:
This backend also uses SQLAlchemy, so we might want to come up with a way to explain the difference between the two. I would expect users of `SQLAlchemy` to use the `flask-sqlalchemy` backend as an example. If someone doesn't use SQLAlchemy and uses another ORM, they might use this as an example of how to use the new data filtering API with their own ORM.

dhatch commented:
Also should add to the README.

dhatch commented on backends/flask/app/__init__.py:
```diff
@@ -0,0 +1,257 @@
+        typ.build_query = fetcher(model)
+        typ.fields = fields
+
+    Relation = import_module("polar.data_filtering", "oso").Relation
```
Curious why this is necessary. Could you just have a top level import? `from polar import data_filtering`?

gw000 replied:
Oh yeah, I guess you can! For some reason I though I couldn't import directly out of `polar`

dhatch commented on backends/flask/app/__init__.py:
```diff
@@ -0,0 +1,257 @@
+    # Initialize SQLAlchemyOso instance.
+    oso = Oso()
+
+    def fetcher(model):
+        def filter(con):
+            if con.field is not None:
+                field = getattr(model, con.field)
+                value = con.value
+            else:
+                field = model.id
+                value = con.value.id
+
+            if con.kind == "Eq":
+                return field == value
+            elif con.kind == "In":
+                return field.in_(value)
+            else:
+                raise OsoError(f"Unsupported constraint kind: {con.kind}")
+
+        def go(cs):
+            pred = and_(*[filter(c) for c in cs])
+            return Session().query(model).filter(pred)
+
+        return go
```
We should use longer variable names here since we expect users to copy this. I put some suggestions below -- didn't test this snippet though. Also, is `fetcher` the name we decided on for these functions? I also thought we were planning on calling constraints filters based on https://www.notion.so/osohq/Fall-2021-Product-Naming-8ff17513d0934d9fbfa3ddfc9dcadb9c#2fe700d5e5dd41bfb22f3830cb1bceb3. Do we need to make updates throughout the code & docs still for this?


```suggestion
    def fetcher(model):
        def to_sqlalchemy_filter(constraint):
            if constraint.field is not None:
                field = getattr(model, constraint.field)
                value = constraint.value
            else:
                field = model.id
                value = constraint.value.id

            if constraint.kind == "Eq":
                return field == value
            elif constraint.kind == "In":
                return field.in_(value)
            else:
                raise OsoError(f"Unsupported constraint kind: {constraint.kind}")

        def build_query(constraints):
            pred = and_(*[to_sqlalchemy_filter(c) for c in constraints])
            return Session().query(model).filter(pred)

        return build_query
```

dhatch replied:
Comments may also be helpful here.

gw000 replied:
👍 added

dhatch reviewed (commented):
Brief comments from reading this code a bit -- not a thorough review.

dhatch commented on backends/flask/app/routes/orgs.py:
```diff
@@ -0,0 +1,38 @@
+def create():
+    payload = request.get_json(force=True)
+    org = Org(**payload)
+    if not current_app.oso.is_allowed(g.current_user, "create", org):
```
Can we use `authorize` here instead?

gw000 replied:
👍

leina05 commented on backends/flask-sqlalchemy/app/routes/helpers.py:
```diff
@@ -4,30 +4,35 @@
-def check_permission(action: str, resource: Base):
-    if not current_app.oso.is_allowed(g.current_user, action, resource):
-        raise Forbidden
+def authorized_resource(actn: str, cls: Type[Any], **kwargs):
```
```suggestion
def authorized_resource(action: str, cls: Type[Any], **kwargs):
```

leina05 commented on backends/flask-sqlalchemy/app/routes/helpers.py:
```diff
@@ -4,30 +4,35 @@
-    if not current_app.oso.is_allowed(g.current_user, action, resource):
-        raise Forbidden
+def authorized_resource(actn: str, cls: Type[Any], **kwargs):
+    rsrcs = authorized_resources(actn, cls, **kwargs)
```
```suggestion
    resources = authorized_resources(action, cls, **kwargs)
```

leina05 commented on backends/flask-sqlalchemy/app/routes/helpers.py:
```diff
@@ -4,30 +4,35 @@
-        raise Forbidden
+def authorized_resource(actn: str, cls: Type[Any], **kwargs):
+    rsrcs = authorized_resources(actn, cls, **kwargs)
+    if len(rsrcs) == 0:
+        raise NotFound
+    return rsrcs[0]
```
```suggestion
    if len(resources) == 0:
        raise NotFound
    return resources[0]
```

leina05 commented on backends/flask-sqlalchemy/app/routes/helpers.py:
```diff
@@ -4,30 +4,35 @@
+    return rsrcs[0]
+
+
+def authorized_resources(actn: str, cls: Type[Any], **kwargs):
```
expand shortened variable names

leina05 commented on backends/flask-sqlalchemy/app/helpers.py:
```diff
@@ -0,0 +1,36 @@
+    return role_name
+
+
+def assign_role(user, resource, role_name, session):
```
Is this used anywhere?

gw000 replied:
No, it's left over from the old backend 👍

leina05 reviewed (changes requested):
The new repo looks great! The old repo needs to have the old `sqlalchemy-roles` stuff cleaned out of it, which I think you could do pretty easily using the implementation you wrote for roles in the new repo. But that can be a follow-up to this PR.

For some reason my test setup is still broken, but others were able to run them and they passed. The `flask-sqlalchemy-oso` repo tests don't pass when running against `oso/main` build because they use the old roles that have now been removed.

One other comment: 
- can we add the new backend to the top-level readme and briefly explain the difference between the two backends?

samscott89 commented on backends/flask-sqlalchemy/app/routes/helpers.py:
```diff
@@ -4,30 +4,35 @@
-        return wrapper
-
-    return decorator
+def session(func):
```
We should use as few of these helpers as possible, so that for someone _not_ using SQLAlchemy it's clear what work they should be doing. We should use regular sessions, and use the data filtering APIs instead.

gw000 replied:
They're all gone 😄

gw000 commented on backends/flask-sqlalchemy/app/models.py:
```diff
@@ -61,7 +61,33 @@ class Issue(Base):
         return {"id": self.id, "title": self.title}
+
+
+class OrgRole(Base):
```
They were created by `sqlalchemy-oso` when we used the built-in RBAC. In the updated `flask-sqlalchemy-oso` they are there.

gw000 commented on backends/flask-sqlalchemy-oso/README.md:
```diff
@@ -0,0 +1,92 @@
+The [Oso documentation][docs] is a good reference for more information on Oso's
+[Python][docs-python] and [SQLAlchemy][docs-sqlalchemy] integrations, and it's
+also where you can find [documentation for the new built-in roles features in
+the `sqlalchemy-oso` library][docs-roles] that this app uses heavily.
```
👍

gw000 commented on backends/flask-sqlalchemy-oso/README.md:
```diff
@@ -0,0 +1,92 @@
+[docs]: https://docs.osohq.com/
+[docs-python]: https://docs.osohq.com/python/reference/lib.html
+[docs-sqlalchemy]: https://docs.osohq.com/python/reference/frameworks/sqlalchemy.html
+[docs-roles]: https://docs.osohq.com/python/new-roles.html
```
✂️

gw000 commented:
Alright everyone thanks for the review, I think I addressed all the issues now. We'll have to update the other backends to make sure they 403 in the right places though.

leina05 reviewed (approved):
🥇
